java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java

```diff
@@ -0,0 +1,237 @@
+    }
+  }
+
+  private static boolean isUndefinedOrNull(CheckerContext context, Symbol symbol) {
+    ProgramState programState = context.getState();
+    SymbolicValue value = programState.getValue(symbol);
+    return value == null;
+  }
+
+  protected String nonNullAnnotation(JavaSymbol javaSymbol) {
+    for (String annotation : ANNOTATIONS) {
+      if (javaSymbol.metadata().isAnnotatedWith(annotation)) {
+        return annotation;
+      }
+    }
+    return null;
+  }
+
+  private class AbstractStatementVisitor extends CheckerTreeNodeVisitor {
+
+    private final CheckerContext context;
+
+    protected AbstractStatementVisitor(CheckerContext context) {
+      super(context.getState());
+      this.context = context;
+    }
+
+    protected void reportIssue(Tree tree, String message, Object... parameters) {
+      context.reportIssue(tree, NonNullSetToNullCheck.this, MessageFormat.format(message, parameters));
+    }
+  }
+
+  private class PreStatementVisitor extends AbstractStatementVisitor {
+
+    protected PreStatementVisitor(CheckerContext context) {
+      super(context);
+    }
+
+    @Override
+    public void visitAssignmentExpression(AssignmentExpressionTree tree) {
+      if (tree.variable().is(Tree.Kind.IDENTIFIER)) {
+        IdentifierTree variable = (IdentifierTree) tree.variable();
+        JavaSymbol symbol = (JavaSymbol) variable.symbol();
+        String nonNullAnnotation = nonNullAnnotation(symbol);
+        if (nonNullAnnotation != null) {
+          List<SymbolicValue> values = programState.peekValues(2);
+          SymbolicValue assignedValue = values.get(1);
+          Constraint constraint = programState.getConstraint(assignedValue);
+          if (constraint != null && constraint.isNull()) {
+            reportIssue(tree, "\"{0}\" is marked \"{1}\" but is set to null.", symbol.name(), nonNullAnnotation);
+          }
+        }
+      }
+    }
+
+    @Override
+    public void visitNewClass(NewClassTree syntaxTree) {
+      Symbol symbol = syntaxTree.constructorSymbol();
+      if (symbol.isMethodSymbol()) {
+        List<SymbolicValue> argumentValues = programState.peekValues(syntaxTree.arguments().size());
+        JavaSymbol.MethodJavaSymbol methodSymbol = (JavaSymbol.MethodJavaSymbol) symbol;
+        checkNullArguments(syntaxTree, methodSymbol.getParameters(),
+          argumentValues, "Parameter {0} to this constructor is marked \"{1}\" but null is passed.");
+      }
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree syntaxTree) {
+      Symbol symbol = syntaxTree.symbol();
+      if (symbol.isMethodSymbol()) {
+        Arguments arguments = syntaxTree.arguments();
+        List<SymbolicValue> argumentValues = new ArrayList<>(programState.peekValues(arguments.size() + 1));
+        argumentValues.remove(0);
+        Collections.reverse(argumentValues);
```

Comment:
I would add a comment to at least distinguish the cases and explain why arguments should be reversed in method invocation when it's not the case for constructors.